source/dofs/dof_renumbering.cc

```diff
@@ -1191,50 +1191,65 @@ namespace DoFRenumbering
   namespace
   {
     // helper function for hierarchical()
-
-// Note that this function only works for active dofs.
-    template <int dim, class iterator>
+    //
+    // Note that this function only works for locally active dofs.
+    template <int dim, class CellIteratorType>
     types::global_dof_index
-    compute_hierarchical_recursive (
-      types::global_dof_index next_free,
-      std::vector<types::global_dof_index> &new_indices,
-      const iterator &cell,
-      const IndexSet &locally_owned)
+    compute_hierarchical_recursive (const types::global_dof_index         next_free_dof_index,
+                                    std::vector<types::global_dof_index> &new_indices,
+                                    const CellIteratorType               &cell,
+                                    const IndexSet                       &locally_owned_dof_indices)
     {
+      types::global_dof_index current_dof_index = next_free_dof_index;
+
       if (cell->has_children())
         {
-          //recursion
+          // recursion
           for (unsigned int c = 0; c < GeometryInfo<dim>::max_children_per_cell; ++c)
-            next_free = compute_hierarchical_recursive<dim> (
-                          next_free,
-                          new_indices,
-                          cell->child (c),
-                          locally_owned);
+            current_dof_index
+              = compute_hierarchical_recursive<dim> (current_dof_index,
+                                                     new_indices,
+                                                     cell->child (c),
+                                                     locally_owned_dof_indices);
         }
       else
         {
           if (cell->is_locally_owned())
             {
+              // first get the existing DoF indices
               const unsigned int dofs_per_cell = cell->get_fe().dofs_per_cell;
               std::vector<types::global_dof_index> local_dof_indices (dofs_per_cell);
               cell->get_dof_indices (local_dof_indices);
 
+              // then loop over the existing DoF indices on this cell
+              // and see whether it has already been re-numbered (it
+              // may have been on a face or vertex to a neighboring
+              // cell that we have encountered before). if not,
+              // give it a new number and store that number in the
+              // output array (new_indices)
+              //
+              // if this is a parallel triangulation and a DoF index is
+              // not locally owned, then don't touch it. since
+              // we don't actually *change* DoF indices (just record new
+              // numbers in an array), we don't need to worry about
+              // the decision whether a DoF is locally owned or not changing
+              // as we progress in renumbering DoFs -- all adjacent cells
+              // will always agree that a DoF is locally owned or not
```

Comment:
I guess the important part here is that the order we traverse the cells matters of course (because we assign dofs increasingly), which is the whole point of this function.

Reply:
Indeed. I've added a sentence to this effect.